init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
Re my last note: our CUDNN experts agree it looks liek a bug in existing code: it would be a tall and unnecessary order to assume the layer to always execute on the same stream - easily fixable with practically no penalty - getSharedWorkspace should always be called and buffer reassigned.